Histogram matching for color calibration of images using segmented color cards

thompsonmj opened the pull request:
This PR does the following:
- Uses RLE segmentation for color cards to generate normalized histograms of only the foreground, eliminating the need to maintain a consistent orientation of the color cards and/or images
- Provides a CLI for bulk processing of an input dataset with option to specify an individual reference image to use as the basis of calibration or default to the first image found with a segmented color card
- Detailed output option showing the before/after images of target color cards and full images as well as an interactive Plotly RGB histogram for each target color card showing the reference channels as well as the target original and transform channels

It currently relies on a `metadata/coco_annotations.json` file to be present in the user-specified argument for the segmented dataset directory, which is the standardized output of `wing-segmentation`. 

This has been tested with the STRI butterfly wing images, though these examples were excluded from this PR since licensing/permissions is uncertain for these images.

Some improvements that should be added as follow-up issues include:
- Improve flexibility of inputs to allow binary masks for color card rather than requiring `wing-segmentation` formatted data
- Examples using permissively licensed images
- Unit tests
- Addition of a calibration method that better accounts for color _perception_ vs just histogram matching
- Parallel calibration

thompsonmj committed:
Repackage using modern pyproject.toml specification

thompsonmj committed:
Update installation and usage instructions

thompsonmj committed:
Implement histogram matching package companion for wingseg

DavidCarlyn commented on src/color_calibration/processor.py:
```diff
@@ -0,0 +1,190 @@
+    """
+    Recursively get all image files in the dataset directory.
+    """
+    image_extensions = {'.jpg', '.jpeg', '.JPG', '.JPEG'}
```
Did we want to allow PNG as well?

egrace479 replied:
Unless it would be problematic to do so, PNGs would be preferable since they're not lossy.

thompsonmj replied:
Will update to allow `Image.open(file_path).verify()` as input.

DavidCarlyn commented:
Other than maybe allowing PNG to also be used as an image format when searching for images, everything looks good to me!

egrace479 commented on README.md:
```diff
@@ -1,14 +1,79 @@
+
+
+## Outputs
+- `--save-dir` will contain the full dataset of calibrated image in the same directory structure as the `--dataset-dir` input dataset. The image used as reference will be copied as-is from the input dataset to keep the full dataset together.
```
```suggestion
- `--save-dir` will contain the full dataset of calibrated images in the same directory structure as the `--dataset-dir` input dataset. The image used as reference will be copied as-is from the input dataset to keep the full dataset together.
```

egrace479 commented on README.md:
```diff
@@ -1,14 +1,79 @@
+
+## Outputs
+- `--save-dir` will contain the full dataset of calibrated image in the same directory structure as the `--dataset-dir` input dataset. The image used as reference will be copied as-is from the input dataset to keep the full dataset together.
+- `calibration_log.json` that tracks the inputs used for the calibration process.
```
```suggestion
- `calibration_log.json` tracks the inputs used for the calibration process.
```

egrace479 commented on README.md:
```diff
@@ -1,14 +1,79 @@
+## Outputs
+- `--save-dir` will contain the full dataset of calibrated image in the same directory structure as the `--dataset-dir` input dataset. The image used as reference will be copied as-is from the input dataset to keep the full dataset together.
+- `calibration_log.json` that tracks the inputs used for the calibration process.
+- If `--detailed-outputs` is enabled, the `--save-dir` will also contain a `detailed-outputs/` directory. This directory will contain:
+    - A directory structure matching the `--dataset-dir` input dataset with the following image types:
+        - `orig_tgt_color_card_<img-name>.png`
+        - `orig_tgt_full_image_<img-name>.png`
+        - `ref_color_card.png`
+        - `ref_full_image_<img-name>.png`
+        - `trans_tgt_color_card_<img-name>.png`
+        - `trans_tgt_full_image_<img-name>.png`
+    - `histograms/` with a directory structure matching the `--dataset-dir` input dataset. The histograms saved as `hist_<img-name>.html` will show the RGB histograms for the reference color card and the target color card before and after transformation for each image. These may be used to interactively inspect the effect of the color calibration on the images.
```
Would it make sense to subdivide the options here or would all extra files be desired if any are? (Just future thoughts)

thompsonmj replied:
Good to consider ... another improvement could be creating an individual figure for each image in the input dataset that contains a subplot for each relevant component like color card before, color card after, full image before, full image after, reference color card, reference full image and RGB histogram. This would make inspection much simpler compared to many images for a given input image under a different prefix.

egrace479 commented on src/color_calibration/__init__.py:
```diff
@@ -1 +1,6 @@
-from .match_histograms import match_histograms
\ No newline at end of file
+__version__ = "0.0.1"
```
```suggestion
__version__ = "0.0.2"
```
This is updated from @DavidCarlyn's original 0.0.1. Though I might argue to at least go to 0.1.0, since it's changing use.

thompsonmj replied:
I think `0.1.0` is sensible since this is intended to be used as part of an external workflow, and it's more than just bugfixes.

egrace479 commented on src/color_calibration/match_histograms.py:
```diff
@@ -1,50 +1,92 @@
 def match_histograms(ref_card, tgt_card, tgt_image):
     """
+    Perform channel-wise histogram matching based on the normalized foreground intensities.
+    No resizing is performed—the normalized cumulative histogram is computed on the raw,
```
```suggestion
    No resizing is performed; the normalized cumulative histogram is computed on the raw,
```
Rendering with the em-dash is a bit ambiguous and a semicolon works here too.

egrace479 commented on src/color_calibration/coco_integration.py:
```diff
@@ -0,0 +1,150 @@
+class DatasetPaths:
+    """Helper class to manage dataset paths and structure"""
+    def __init__(self, dataset_base_dir, dataset_output_dir):
+        self.base_dir = Path(dataset_base_dir)
+        self.output_dir = Path(dataset_output_dir)
+        self.resized_dir = self.output_dir / "resized"
+        
+        if not self.base_dir.exists():
+            raise FileNotFoundError(f"Dataset base directory not found: {self.base_dir}")
+        if not self.output_dir.exists():
+            raise FileNotFoundError(f"Dataset output directory not found: {self.output_dir}")
```
```suggestion
        self.base_dir = Path(dataset_base_dir)
        if not self.base_dir.exists():
            raise FileNotFoundError(f"Dataset base directory not found: {self.base_dir}")
        self.output_dir = Path(dataset_output_dir)
        if not self.output_dir.exists():
            raise FileNotFoundError(f"Dataset output directory not found: {self.output_dir}")
        
        self.resized_dir = self.output_dir / "resized"
```
Should save the definition of resized directory to after confirmation that the output directory exists.

egrace479 reviewed (commented):
A few suggestions for clarity.

thompsonmj committed:
Apply suggestions from code review

Improve some wording; fix spelling errors; adjust order of DatasetPaths attributes and validations

Co-authored-by: Elizabeth Campolongo <38985481+egrace479@users.noreply.github.com>

thompsonmj committed:
Version increment; allow any input image format

thompsonmj committed:
Merge branch 'main' into feat/segmentation-masks
